[Enhancement] Security & Auto-approve answers by instructors

filip-gregurevic opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
An answer given by an instructor is not highlighted except if it is approved manually.
There are insufficient checks for security with the questions and answers.

### Description
<!-- Describe your changes in detail -->
Answers given by an instructor are now approved automatically.
Include course in api calls and check the user's access right in all endpoints.


### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to an exercise or lecture with a question
3. Write answers with different access rights (e.g student, instructor)
4. Check that only answers given by instructors are approved automatically.

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
* StudentQuestionAnswerResource.java: 92%

JonasPetry reviewed (approved):
Incredible 🚀

stefanwaldhauser reviewed (dismissed):
Good Job! Works as expected

MaisiKoleni commented on src/test/java/de/tum/in/www1/artemis/StudentQuestionAnswerIntegrationTest.java:
```diff
@@ -67,10 +67,44 @@ public void createStudentQuestionAnswer() throws Exception {
         StudentQuestionAnswer response = request.postWithResponseBody("/api/student-question-answers", studentQuestionAnswer, StudentQuestionAnswer.class, HttpStatus.CREATED);
 
+        // should be automatically approved
+        assertThat(response.isTutorApproved());
```
```suggestion
        assertThat(response.isTutorApproved()).isTrue();
```

MaisiKoleni commented on src/test/java/de/tum/in/www1/artemis/StudentQuestionAnswerIntegrationTest.java:
```diff
@@ -67,10 +67,44 @@ public void createStudentQuestionAnswer() throws Exception {
+        StudentQuestionAnswer response = request.postWithResponseBody("/api/student-question-answers", studentQuestionAnswer, StudentQuestionAnswer.class, HttpStatus.CREATED);
+
+        // shouldn't be automatically approved
+        assertThat(!response.isTutorApproved());
```
```suggestion
        assertThat(response.isTutorApproved()).isFalse();
```

MaisiKoleni commented on src/test/java/de/tum/in/www1/artemis/StudentQuestionAnswerIntegrationTest.java:
```diff
@@ -67,10 +67,44 @@ public void createStudentQuestionAnswer() throws Exception {
+        StudentQuestionAnswer response = request.postWithResponseBody("/api/student-question-answers", studentQuestionAnswer, StudentQuestionAnswer.class, HttpStatus.CREATED);
+
+        // shouldn't be automatically approved
+        assertThat(!response.isTutorApproved());
```
```suggestion
        assertThat(response.isTutorApproved()).isFalse();
```

stefanwaldhauser reviewed (approved):
reapprove

MaisiKoleni reviewed (dismissed):
Works well and code looks good to me now.

anditurdiu reviewed (dismissed):
Works as expected!

fde312 reviewed (changes requested):
Getting a internal server error, for adding answers for this exercise:
https://artemistest.ase.in.tum.de/#/courses/256/exercises/4109
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/41108487/95121173-f4c00300-074e-11eb-9b65-9fb94e52e61f.png">

fde312 reviewed (dismissed):
Works now! Great job 🚀

sjagla27 reviewed (dismissed):
Works as expected.

balazs-czopf reviewed (dismissed):
Works as expected. Nice work 🚀

schultek reviewed (dismissed):
Works now as expected. Great job.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:
```diff
@@ -61,19 +66,28 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
-    // TODO: there are no security checks here. The API endpoint should at least include the course id
-    public ResponseEntity<StudentQuestionAnswer> createStudentQuestionAnswer(@RequestBody StudentQuestionAnswer studentQuestionAnswer) throws URISyntaxException {
+    public ResponseEntity<StudentQuestionAnswer> createStudentQuestionAnswer(@PathVariable Long courseId, @RequestBody StudentQuestionAnswer studentQuestionAnswer)
+            throws URISyntaxException {
```
there is still no security check here. Please check that the user is at least a student in the course

krusche replied:
If you touch this file, there are additional TODOs regarding the REST URLs and the security checks. Could you please fix those for all requests in this class and in the class StudentQuestionResource, so that all URLs and security checks are correct and consistent?
Thank you!

krusche reviewed (changes requested):
I added one inline comment regarding REST URLs and security checks, please fix those issue in all places for StudentQuestions and StudentQuestionAnswers, thanks!

filip-gregurevic commented:
> I added one inline comment regarding REST URLs and security checks, please fix those issue in all places for StudentQuestions and StudentQuestionAnswers, thanks!

@krusche I now also fixed all todos for those two. The course access is now checked everywhere where it wasn't before.

fde312 reviewed (changes requested):
When I give a answer as a student, it gets also auto approved:

<img width="540" alt="image" src="https://user-images.githubusercontent.com/41108487/95690879-263b4180-0c1b-11eb-942f-5123d1b0cdae.png">

stefanwaldhauser commented:
> When I give a answer as a student, it gets also auto approved:
> 
> <img alt="image" width="540" src="https://user-images.githubusercontent.com/41108487/95690879-263b4180-0c1b-11eb-942f-5123d1b0cdae.png">

![image](https://user-images.githubusercontent.com/29718932/95690917-6b5f7380-0c1b-11eb-8125-9c54c78cd112.png)
I think that is a expected as TS1 is a admin for some reason on TS1

fde312 commented:
> > When I give a answer as a student, it gets also auto approved:
> > <img alt="image" width="540" src="https://user-images.githubusercontent.com/41108487/95690879-263b4180-0c1b-11eb-942f-5123d1b0cdae.png">
> 
> ![image](https://user-images.githubusercontent.com/29718932/95690917-6b5f7380-0c1b-11eb-8125-9c54c78cd112.png)
> I think that is a expected as TS1 is a admin for some reason on TS1

Jupe, the issue is probably because test user 1 and test user 2 have also the role admin/instructor. For Test user 3 (which does not have admin/instructor role) it works fine:
<img width="510" alt="image" src="https://user-images.githubusercontent.com/41108487/95691061-8da5c100-0c1c-11eb-9e90-4e6b4e93ffe4.png">

@filip-gregurevic we probably need to check the roles for the specific course.

stefanwaldhauser reviewed (dismissed):
Looked through the code and everything works as expected. Great job :)

tiffca reviewed (approved):
Works well, answers from tutors and instructors are auto-approved as expected. 

Steps (test server 1) 
Created a question as a student (test user 4). 
Answered question as tutor (test user 6). - auto approved
Answered question as instructor (test user 8). - auto approved
Answered question as student (test user 4).  
Answered question as student (test user 10). 

![image](https://user-images.githubusercontent.com/63906424/95765494-3eaa6b00-0cb2-11eb-9c3d-71c8bcdf0997.png  | width=100))

fde312 commented:
> Works well, answers from tutors and instructors are auto-approved as expected.
> 
> Steps (test server 1)
> Created a question as a student (test user 4).
> Answered question as tutor (test user 6). - auto approved
> Answered question as instructor (test user 8). - auto approved
> Answered question as student (test user 4).
> Answered question as student (test user 10).
> 
> <img alt="" width="40%" src="https://user-images.githubusercontent.com/63906424/95765494-3eaa6b00-0cb2-11eb-9c3d-71c8bcdf0997.png">
I think answers from tutors should not be auto-approved? 😅

tiffca reviewed (changes requested):
I misunderstood the original requirement, changing my review to request changes per Francisco's comment. I thought tutor answers were also supposed to be auto-approved because the hover over the answer mentions "approved by a tutor".

filip-gregurevic commented:
> I misunderstood the original requirement, changing my review to request changes per Francisco's comment. I thought tutor answers were also supposed to be auto-approved because the hover over the answer mentions "approved by a tutor".

@tiffca Which course did you use to test? The roles of each test user are not really up-to-date anymore and you have to check for every course

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:
```diff
@@ -50,30 +53,43 @@
+            return forbidden();
+        }
+        // answer to approved if written by an instructor
+        studentQuestionAnswer.setTutorApproved(this.authorizationCheckService.isAtLeastInstructorInCourse(optionalCourse.get(), user));
```
As a malicious API user, I can circumvent this check by providing a course id not matching the `studentQuestionAnswer`. In this case, I can auto-approve messages while only being an instructor in a different course.

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:
```diff
@@ -83,27 +99,32 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
         if (studentQuestionAnswer.getId() == null) {
             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
         }
+        Optional<Course> optionalCourse = courseRepository.findById(courseId);
+        if (optionalCourse.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
```
Similar to above, this does not check if `studentQuestionAnswer` is in the same course as `courseId`.

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:
```diff
@@ -118,35 +139,47 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
+    public ResponseEntity<StudentQuestionAnswer> getStudentQuestionAnswer(@PathVariable Long courseId, @PathVariable Long id) {
         log.debug("REST request to get StudentQuestionAnswer : {}", id);
+        User user = this.userService.getUserWithGroupsAndAuthorities();
+        Optional<Course> optionalCourse = courseRepository.findById(courseId);
+        if (optionalCourse.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
+        if (!this.authorizationCheckService.isAtLeastStudentInCourse(optionalCourse.get(), user)) {
+            return forbidden();
+        }
```
Similar to above, this does not check if the requested question is in the same course as `courseId`.

jpbernius commented on src/main/java/de/tum/in/www1/artemis/web/rest/StudentQuestionAnswerResource.java:
```diff
@@ -118,35 +139,47 @@ public StudentQuestionAnswerResource(StudentQuestionAnswerRepository studentQues
         if (optionalStudentQuestionAnswer.isEmpty()) {
             return ResponseEntity.notFound().build();
         }
+        Optional<Course> optionalCourse = courseRepository.findById(courseId);
+        if (optionalCourse.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
```
Similar to above, this does not check if the requested question is in the same course as `courseId`.

jpbernius reviewed (changes requested):
This Feedback applies to all resources:

krusche commented:
As this PR already improves the security related to quiz questions, I would merge it now. The mentioned issues found by @jpbernius can be fixed in a smaller follow-up PR.

@filip-gregurevic please take care about solving those issues in a follow-up. Thanks!
